docs/design/data_safety.md

```diff
@@ -0,0 +1,153 @@
+Data Safety in ICU4X
+====================
+
+At its core, ICU4X is a set of algorithms that map input data to human-ready outputs. Since a core value proposition of ICU4X is the ability to dynamically load data, we can't always assume that our data are valid and pristine. This doc covers how we think about this problem and handle it in ICU4X.
+
+## Resolution
+
+1. _Whereas_ it is rare to be 100% confident about the safety of your data\*; and
+1. _Whereas_ validating data invariants can be expensive; and
+1. _Whereas_ moving validation into the algorithm (post-deserialization) often reduces the overall overhead of validation; and
```

Comment:
I think this fails to address the concern of not having validation overhead at all when data provenance and integrity has been assured via `BakedDataProvider`. To avoid overhead for users of `BakedDataProvider` it's preferable to validate data at deserialization (and then omit that step with `BakedDataProvider`) as opposed to moving validation into the algorithm.

Chances are that the deserialization-time validation cost is more likely to be acceptable for simple Unicode Database data than for complex CLDR data.

Reply:
I revised the text and added a new point. PTAL.

Reply:
Looking at the new text now, it's unclear to me if policy would prohibit or merely seemingly discourage stuff like checking at deserialization time that things meant to be scalar values indeed are. By precedent from validating `&str`, it seems we should be OK with validating the values of e.g. the normalization trie for scalar range as applicable, but it seems to me I can't find support from that from this text as drafted.

Reply:
https://github.com/unicode-org/icu4x/pull/2768/commits/b7116f193df02fd9e4ad0ed35afbc641069ce266